Features/eval structure

kuraisle opened the pull request:
## Pull Request Contents

<!--
ACTION: Delete any content types that don't apply to your pull request
-->

| |
|-|
✨ Feature

## Description

<!--
ACTION: Summarise your Pull Request
-->
I'm trying to plan how Llettuce evaluation can be carried out. It needs to be reproducible, so anyone can test the pipelines and compare their results to past runs.

I think defining a class is a good first point. It should do two things:
1. Run evaluations
2. Save the results

```python
class EvaluationFramework:
	def __init__(self, results_file='results.json'):
		self.results_file = results_file
	def run_evaluations(self):
		# Run some tests
		self._save_evaluations
	def _save_evaluations(self):
		# Append to 'results.json'
```

Of course, it needs to run evaluations. What is an evaluation? It's running some sort of metric on the results of a pipeline run. We can put this in a class too. We'll be using [[Abstract base class|ABCs]] here.
A Metric just has to calculate a result
```python
class Metric(ABC):
    """Base class for all metrics."""
    @abstractmethod
    def calculate(self, *args, **kwargs):
        """
        Calculate the metric value.
        """
        pass
```
A TestPipeline has to run
```python
class TestPipeline(ABC):
	"""
	Base class for Pipeline runs
	"""
	@abstractmethod
	def run(self, *args, **kwargs):
		"""
		Run the pipeline
		"""
		pass
```
A PipelineTest has to run the pipeline, then calculate any metrics
```python
class PipelineTest(ABC):
	"""
	Base class for Pipeline tests
	"""
	def __init__(self, name: str, pipeline: TestPipeline, metrics: list[Metric]):
		self.pipeline = pipeline
		self.metrics = metrics
	@abstractmethod
	def run_pipeline(self, *args, **kwargs):
		pass
	def evaluate(self, *args, **kwargs):
		self.performance = [metric.run(self.pipeline_result) for metric in metrics]
```
Once we have these classes, we can start subclassing
```python
class SingleResultMetric(Metric):
    """Metric for evaluating pipelines that return a single result."""
    pass

class InformationRetrievalMetric(Metric):
    """Metric for evaluating information retrieval pipelines."""
    pass
```
These classes can make sure we only have the right metrics for our pipelines

## Related Issues or other material

- [X] This PR relates to one or more issues, detailed below

<!--
ACTION: Detail any issues this PR relates to, and then check the box.

NOTE: PRs without issues will not be merged.
-->

<!--
We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.

Be careful not to close parent tracking issues that this PR only completes PART of.
-->

- Related Issue #16 
- Closes #16 

## Screenshots, example outputs/behaviour etc.

<!--
ACTION: Add material or delete this section if it's not applicable
-->

## ✅ Added/updated tests?

- [X] This PR contains relevant tests
  - Or doesn't need to per the below explanation

<!--
ACTION: If tests have not been added or updated, explain why here
-->
Tests will only be necessary when we start running evaluations


- [X] I've completed all **actions** and **tasks** detailed in the PR Template

kuraisle committed:
Updated requirements

The version of haystack we were locked on has a security risk, so I've updated the requirements and environment.yml to the latest version for testing

kuraisle committed:
Added starlette to conda environment

kuraisle committed:
Added basics of evaluation framework

CodeByKarthik reviewed (changes requested):
Greetings James, 

The evaluation framework is good and the code structure is reusable and scalable. In future, when we are developing, we can adjust the code as necessary. 

But I noted one small issue which I've mentioned below:
- There is a conflict in the branch on requirement.txt. 

Other than that, there are no issues. Good Job 👍

kuraisle committed:
Updated requirement

branched from a version where the requirements.txt and environment.yml weren't fixed

kuraisle committed:
Update environment.yml

kuraisle commented:
> Greetings James,
> 
> The evaluation framework is good and the code structure is reusable and scalable. In future, when we are developing, we can adjust the code as necessary.
> 
> But I noted one small issue which I've mentioned below:
> 
>     * There is a conflict in the branch on requirement.txt.
> 
> 
> Other than that, there are no issues. Good Job 👍

I branched from a version without the right requirements. Fixed now. If the previous PR is merged first, it shouldn't show up now

CodeByKarthik committed:
Merge branch 'main' into features/eval-structure

Karthi-DStech commented:
I've resolved the conflict and merged to the main.
